### ISSUE TITLE
Parts of reconciliation batches can fail silently; and parts of reconciled column lookup batches can fail silently

### ISSUE THREAD

**Issue by Jheald** [OpenRefine/OpenRefine#3369]
This may be a duplicate of an existing bug or bugs, but I find it incredibly frustrating: when items are 'unmatched' it appears to be impossible to distinguish whether there is really no match for the item, or whether the reconciliation process simply failed to complete for some sub-batch of entries.

A couple of days ago, I was running a reconciliation of a set of authors with LoC authority IDs against Wikidata using the "reconcili.link" service.  I ran it once, and about 80% reconciled.  I then filtered for the unmatched authors, reconciled again, and a further 15% of the original set reconciled.

I presume that what happened was that during the first reconciliation, one or more sub-batches of contiguous records timed out, or otherwise failed to return.  (The problem did appear to be affecting contiguous records).   That, I guess, is something that can always happen.  The problem is that OpenRefine then marked those records' reconciliation status as "unmatched" rather than "unknown", not differentiating between items for which the reconcilation had failed to complete from those for which the reconcilation had concluded but found nothing.

Similarly, when I then added a column looking up LoC IDs based on the reconciled values (which I find is a necessary thing to have to do to sanity check reconcilations), only about 80% of the IDs got added, with more added when I re-ran the lookup for those without values.

This is frustrating enough for me, aware that this can happen, knowing that I may need to re-run reconciliations or augmentations two or three or even more times to be sure.  But the even worse impact is with unaware users, who as a result may then add new duplicate items to Wikidata for their unreconciled entries, without realising that the unreconciled status cannot be depended on.

**Comment by thadguidry** [comment]
@Jheald You didn't provide any details from our bug template such as OpenRefine version, JDK version, OS, etc.  Please provide those.

**Comment by Jheald** [comment]
Can't find your bug template now (another bug?) but:

OS:  Windows 10 Home, version 1909  build 18363.1198  running on Intel 64-bit
Java version 1.8.0_231
Jave SE runtime environment build 1.8.0_231-b11
Java 64-bit server VM build 25.231-b11
OpenRefine version 3.4.1 [437dc4d]

but I think I have also encountered the same problem with earlier versions of OpenRefine

**Comment by tfmorris** [comment]
This is a long standing problem made worse by the fact that the originally designed retry mechanism is broken, so you only get one shot per item.

https://github.com/OpenRefine/OpenRefine/blob/96789f994c8f61af4c07e1b6bc79d369784f09df/main/src/com/google/refine/model/recon/StandardReconConfig.java#L526

It's not an excuse, but when the reconciliation code was designed, the primary reconciliation was very reliable and high performance. Today's reconciliation services -- not so much. A 15% failure rate is outrageous! You might consider also asking the operator of your favorite reconciliation service to strive for more like <1% -- not that that means we shouldn't fix OpenRefine's reconciliation. Patches gratefully accepted.

**Comment by wetneb** [comment]
I will try to have a go at this soon, it would be good to have this fixed for 3.5.

**Comment by tfmorris** [comment]
Do retries fully resolve this? The scenario that I'm thinking of is reconciliation candidate that a service has hard failure with for whatever reason (reconciliation service bug, most likely). I don't think the user can detect this case and, I'm not sure, but I think it might kill the entire batch that the bad candidate is in.

Certainly for unreliable reconciliation services, the retries will provide a vast improvement.

**Comment by wetneb** [comment]
Yes it's a good point: the PR does not address the issue fully, we should have a way to signal to the user that some batches failed. It's not clear to me how we should do this. Perhaps simply replace the corresponding cell values by an `EvalError`? Currently, reconciliation doesn't touch the cell values at all, so that might be something we want to preserve, but I cannot really think of other options.

**Comment by wetneb** [comment]
I think using `EvalError` could work for this, I'll try to get a PR in, hopefully that could make it into 3.5.

**Comment by thadguidry** [comment]
Totally agree with @Jheald that the most important thing firstly is giving good feedback on what happened to our users.  Who can then rerun however they wish, even using different reconcile services if they so desire or have an expectation that one might perform or be better than another.
